src/plugin.ts

```diff
@@ -0,0 +1,252 @@
+    buildFilePathWithPrefix({ itemId, pathPrefix, filename });
+
+  const createThumbnails = async (item: Item<UnknownExtra>, actor: Actor) => {
+    // get original image
+    const filename = getFilePathFromItemExtra(
+      serviceMethod,
+      item.extra as FileItemExtra,
+    );
+    const task = fileTaskManager.createGetFileBufferTask(actor, { filename });
+    const originalImage = await runner.runSingle(task);
+
+    // generate sizes for given image
+    const files = THUMBNAIL_SIZES.map(({ name, width }) => ({
+      size: name,
+      image: sharp(originalImage).resize({ width }).toFormat(THUMBNAIL_FORMAT),
+    }));
+
+    return files;
+  };
+
+  fastify.register(basePlugin, {
+    serviceMethod, // S3 or local
+    buildFilePath: buildFilePath,
+    serviceOptions,
+
+    // use function as pre/post hook to avoid infinite loop with thumbnails
+    uploadPreHookTasks: (data, auth) => {
+      // check file is an image
+      if (!data.mimetype.includes('image')) {
+        throw new Error('File is not an image');
+      }
+
+      return options?.uploadPreHookTasks?.(data, auth);
+    },
+
+    uploadPostHookTasks: async ({ file, itemId }, { member }) => {
+      const thumbnails = THUMBNAIL_SIZES.map(({ name, width }) => ({
+        size: name,
+        image: sharp(file).resize({ width }).toFormat(THUMBNAIL_FORMAT),
+      }));
+
+      // it might not be saved correctly in the original upload
+      return await Promise.all(
+        thumbnails.map(async ({ size: filename, image }) =>
+          fileTaskManager.createUploadFileTask(member, {
+            file: await image.toBuffer(),
+            filepath: buildFilePath(itemId, filename),
+            mimetype: THUMBNAIL_MIMETYPE,
+          }),
+        ),
+      );
+    },
+    downloadPreHookTasks: options.downloadPreHookTasks,
+  });
+
+  if (enableItemsHooks) {
+    const deleteFileTaskName = itemTaskManager.getDeleteTaskName();
+    runner.setTaskPostHookHandler<Item>(
+      deleteFileTaskName,
+      async ({ id }, actor, { log = defaultLogger }) => {
+        //  check item has thumbnails
+        try {
+          // await access(buildFilePath(id, undefined))
+          // delete thumbnails for item
+          const tasks = THUMBNAIL_SIZES.map(({ name }) => {
+            const filepath = buildFilePath(id, name);
+            return fileTaskManager.createDeleteFileTask(actor, {
+              filepath,
+            });
+          });
+          // no need to wait for thumbnails to be deleted
+          runner.runMultiple(tasks, log);
+        } catch (err) {
+          log.error(err);
+        }
+      },
+    );
+
+    const copyItemTaskName = itemTaskManager.getCopyTaskName();
+    runner.setTaskPostHookHandler<Item>(
+      copyItemTaskName,
+      async (item, actor, { log = defaultLogger }, { original }) => {
+        const { id } = item; // full copy with new `id`
+        try {
+          // copy thumbnails for copied item
```

Comment:
and check if thumbnails exist as above.
Eventually we can create a separate issue for that.

Reply:
I'm gonna do it in a separate issue